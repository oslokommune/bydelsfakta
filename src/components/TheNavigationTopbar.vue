<template>
  <header class="oslo__navigation-topbar" @keydown.esc="showDropdown = false">
    <h1 v-if="selectedTopic === null" id="select" class="header">{{ getDistrict($route.params.district) }}</h1>
    <div class="navigation-topbar" :class="{ 'navigation-topbar--hidden': selectedTopic === null }">
      <div
        tabindex="0"
        role="button"
        v-if="selectedTopic"
        id="select"
        class="navigation-topbar__select"
        @click="showDropdown = !showDropdown"
        @keydown.enter="showDropdown = !showDropdown"
        @keydown.escape="closeMenu"
        v-click-outside="closeMenu"
        ref="select"
        :aria-label="$t('navigationTopbar.selectTopic.aria')"
        :title="$t('navigationTopbar.selectTopic.aria')"
      >
        <div class="label" :class="{ 'label--active': selectedTopic !== null }">
          {{ $t('navigationTopbar.selectTopic.label') }}
        </div>
        <div class="topic">
          <h1 class="header">{{ getHumanReadableTopic(selectedTopic) }}</h1>
          <ok-icon icon-ref="arrow-down" :options="{ rotation: showDropdown ? '180deg' : false }"></ok-icon>
        </div>
      </div>
      <transition name="fade">
        <div id="dropdown" class="navigation-topbar__dropdown" v-if="showDropdown">
          <div v-for="(kategori, index) in dropdown" :key="index" class="navigation-topbar__dropdown-column">
            <div
              class="navigation-topbar__dropdown-column--heading"
              :style="{ color: kategori.color, 'border-top-color': kategori.color }"
            >
              <span>{{ kategori.kategori }}</span>
            </div>
            <div class="navigation-topbar__dropdown-items">
              <router-link
                v-for="(link, topicIndex) in kategori.links"
                class="navigation-topbar__dropdown-item"
                :id="`dropdown-href-${topics[link].value}`"
                :class="{
                  'navigation-topbar__dropdown-item--active': checkActiveTopic(topics[link].value),
                  'navigation-topbar__dropdown-item--disabled': disabledTopics.includes(topics[link].value),
                }"
                :key="topicIndex"
                v-text="topics[link].text"
                :to="onClickTopic(topics[link].value)"
                >{{ topics[`${link}`].text }}</router-link
              >
            </div>
          </div>
        </div>
      </transition>
    </div>
  </header>
</template>
<script>
import { mapState, mapActions } from 'vuex';
import allDistricts from '../config/allDistricts';
import { categories, topics, disabledTopics } from '../config/topics';
import OkIcon from './OkIcon';

export default {
  name: 'TheNavigationTopbar',
  data() {
    return {
      selectedTopic: null,
      dropdown: categories,
      allDistricts: allDistricts,
      showDropdown: false,
      topics: topics,
      disabledTopics: disabledTopics,
    };
  },

  components: { OkIcon },

  computed: {
    ...mapState(['compareDistricts', 'menuIsOpen', 'districts']),
  },

  methods: {
    ...mapActions(['setMenuIsOpen']),

    closeMenu() {
      if (this.showDropdown) {
        this.showDropdown = false;
      }
    },

    getHumanReadableTopic(id) {
      if (this.topics[id]) {
        return this.topics[id].text;
      } else {
        this.selectedTopic = null;
        return '';
      }
    },

    getDistrict(id) {
      if (this.compareDistricts || id === 'alle') {
        return this.$t('navigationTopbar.header.compareDistrict');
      } else if (this.$route.name === 'NotFound') {
        return this.$t('navigationTopbar.header.notFound');
      } else if (this.$route.name === 'Home') {
        return this.$t('navigationTopbar.header.home');
      } else {
        return id !== undefined
          ? this.allDistricts.find(district => district.uri === id).value
          : this.$t('navigationTopbar.header.chooseDistrict');
      }
    },

    checkActiveTopic(topic) {
      return this.$route.path.includes(topic);
    },

    onClickTopic(topic) {
      return this.disabledTopics.includes(topic)
        ? ''
        : { name: 'Topic', params: { district: this.$route.params.district, topic } };
    },
  },

  watch: {
    $route(to) {
      const routes = to.path.split('/');

      if (to.name !== 'Topic') {
        this.selectedTopic = null;
        this.showDropdown = false;
      } else {
        if (routes.length > 3) this.selectedTopic = routes[3];
      }
      this.closeMenu();
    },

    showDropdown() {
      this.setMenuIsOpen(this.showDropdown);
    },
  },
};
</script>
<style scoped lang="scss">
@import '../styles/colors';
@import '../styles/variables';

.header {
  color: $color-purple;
  display: none;
  flex-direction: column;
  font-size: $font-huge;
  font-weight: 500;
  margin-left: 0.5rem;
  padding: 0.5rem;
  position: relative;
  width: auto;

  @media screen and (min-width: $break-lg) {
    display: block;
  }
}

.oslo__navigation-topbar {
  background-color: $color-bg;
  display: flex;
  flex-direction: column;
  margin: 1.5rem 0.5rem 0.5rem 0.5rem;
  padding: 1rem 0;
  position: relative;
  top: 0;
  width: calc(100% - 1rem);
  z-index: 3;

  @media screen and (min-width: $break-lg) {
    margin: 0;
    padding: 0.75rem 1rem 0.5rem;
    width: 100%;
  }

  &-button {
    display: none;
    flex-direction: row;

    @media screen and (min-width: $break-lg) {
      display: flex;
    }
  }
}

.oslo__topbar {
  color: $color-purple;
  font-weight: bold;
  margin-bottom: 2rem;

  &-text {
    letter-spacing: 0.8px;
    margin-left: 1rem;
    margin-top: 0.3rem;
    text-transform: uppercase;
  }
}

.label {
  font-size: $font-medium;
  margin-bottom: 0.25rem;
}

.topic {
  align-items: center;
  display: flex;
  text-align: left;
  width: 100%;

  .header {
    display: block;
    margin: 0;
    padding: 0;
    width: auto;
  }

  & > .icon {
    margin-left: 0.5rem;
  }
}

.icon {
  background-position: 50% 50%;
  background-repeat: no-repeat;
  height: 2rem;
  transition: all 0.5s cubic-bezier(0.3, 0, 0.5, 1);
  width: 2rem;

  &.rotate {
    transform: rotateX(180deg);
  }

  &__arrow {
    background-image: url('/icons/arrow.svg');
  }
}

.navigation-topbar {
  max-width: 768px;
  position: relative;
  width: 100%;

  &--hidden {
    @media screen and (max-width: $break-lg) {
      display: none;
    }
  }

  &__select {
    align-items: flex-start;
    color: $color-purple;
    display: flex;
    flex-direction: column;
    font-size: $font-huge;
    font-weight: 500;
    padding: 0.5rem;
    position: relative;
    width: 100%;

    @media screen and (min-width: $break-md) {
      padding: 1rem;
    }

    &:hover {
      background-color: darken($color-bg, 5%);
      cursor: pointer;
    }
  }

  &__dropdown {
    background-color: white;
    border: 3px solid $color-border;
    display: flex;
    flex-flow: row wrap;
    max-height: calc(100vh - 17rem);
    max-width: 768px;
    overflow: auto;
    padding: 1rem 0.5rem;
    position: absolute;
    width: 100%;
    z-index: 1;

    &-column {
      display: flex;
      flex: 33.333%;
      flex-direction: column;
      min-width: 200px;
      padding: 0.5rem;

      &--heading {
        font-weight: 700;
        margin-bottom: 1rem;
        margin-top: 1rem;
        padding-top: 0.5rem;

        & > span {
          padding-left: 0.5rem;
        }

        @media screen and (min-width: $break-lg) {
          margin-top: 0;
        }
      }
    }

    &-items {
      display: flex;
      flex-direction: column;
    }

    &-item {
      background-color: $color-grey-50;
      border-radius: 4px;
      color: $color-purple;
      margin-bottom: 0.5rem;
      padding: 1rem 0.75rem;

      @media screen and (min-width: $break-lg) {
        margin-bottom: 0.1rem;
        padding: 0.5rem;
      }

      &:hover:not(&--active):not(&--disabled) {
        background-color: $color-grey-100;
        color: black;
      }

      &--active {
        background-color: $color-blue;
      }

      &--disabled {
        background-color: white;
        opacity: 0.45;
      }
    }
  }
}
</style>
