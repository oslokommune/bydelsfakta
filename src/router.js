import Vue from 'vue';
import Router from 'vue-router';

import { topicNames, disabledTopics } from './config/topics';
import allDistricts from './config/allDistricts';

const District = () => import('./views/District');
const Topic = () => import('./views/Topic');
const NotFound = () => import('./views/NotFound');

Vue.use(Router);

export const routes = [
  {
    path: '/',
    name: 'Home',
    component: District,
  },
  {
    path: '/bydel/:district?',
    name: 'District',
    component: District,
    props: true,
  },
  {
    path: '/bydel/:district/:topic',
    name: 'Topic',
    component: Topic,
    props: true,
  },
  {
    path: '*',
    name: 'NotFound',
    component: NotFound,
  },
];

const router = new Router({
  mode: 'history',
  routes,
});

// Temp while we find out what our homepage should look like
router.beforeEach((to, from, next) => {
  if (to.params.topic) {
    if (!topicNames.find(name => name === to.params.topic) || disabledTopics.includes(to.params.topic)) {
      next({ name: 'NotFound', params: [to.path] });
    }
  }

  if (to.params.district) {
    const districts = to.params.district.split('-');
    if (districts[0] === 'alle') {
      next();
    } else if (districts.length === 1) {
      if (allDistricts.find(district => district.uri === districts[0])) {
        next();
      } else if (allDistricts.find(district => district.key === districts[0])) {
        next();
      } else {
        next({ name: 'NotFound', params: [to.path] });
      }
    } else if (districts.length > 1) {
      const errors = [];
      districts.forEach(district => {
        if (!allDistricts.find(item => item.key === district)) {
          errors.push(district);
        }
      });
      if (errors.length > 0) {
        next({ name: 'NotFound', params: [to.path] });
      }
    }
  }

  if (to.params.topic !== from.params.topic || (to.name === 'Topic' && from.name === 'District')) {
    window.scroll(0, 0);
  }

  next();
});

export default router;
