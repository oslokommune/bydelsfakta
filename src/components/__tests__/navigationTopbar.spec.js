import 'whatwg-fetch';
import { createLocalVue, mount } from '@vue/test-utils';
import Vuex from 'vuex';
import TheNavigationTopbar from '../TheNavigationTopbar';
import router from '../../router';
import clickOutside from '../../directives/clickOutside';
import store from '../../store';
import setupI18n from '../../i18n';

const i18n = setupI18n();

global.scroll = jest.fn();
window.scroll = jest.fn();

describe('TheNavigationTopbar', () => {
  let wrapper = null;

  beforeEach(() => {
    const localVue = createLocalVue();
    localVue.use(router);
    localVue.use(Vuex);
    localVue.directive('click-outside', clickOutside);
    wrapper = mount(TheNavigationTopbar, {
      localVue,
      router,
      store,
      i18n,
      stubs: {
        'ok-icon': true,
      },
    });
  });

  afterEach(() => {
    wrapper.destroy();
  });

  test('renders navigationDrawer-component and finds oslo__navigation-topbar-class', () => {
    expect(wrapper.classes('oslo__navigation-topbar')).toBe(true);
  });

  test('renders correctly', () => {
    router.push('/bydel/sagene');
    expect(wrapper.element).toMatchSnapshot();
  });

  test('return false if subpage is not active', () => {
    router.push('/bydel/sagene/boligpriser');
    expect(wrapper.vm.checkActiveTopic('levekaar')).toEqual(false);
  });

  test('change showDropdown to false if it is true', () => {
    router.push('/bydel/sagene/boligpriser');
    wrapper.setData({ showDropdown: true });
    wrapper.vm.closeMenu();
    expect(wrapper.vm.showDropdown).toEqual(false);
  });

  test('keep showDropdown as false if false', () => {
    router.push('/bydel/sagene/boligpriser');
    wrapper.vm.closeMenu();
    expect(wrapper.vm.showDropdown).toEqual(false);
  });

  test('return router object when clicking on a subpage', () => {
    router.push('/bydel/sagene/boligpriser');
    expect(wrapper.vm.onClickTopic('boligpriser')).toEqual({
      name: 'Topic',
      params: { district: 'sagene', topic: 'boligpriser' },
    });
  });
});
