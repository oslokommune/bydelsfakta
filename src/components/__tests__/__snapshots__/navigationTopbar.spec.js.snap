// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TheNavigationTopbar renders correctly 1`] = `
<header
  class="oslo__navigation-topbar"
>
  <h1
    class="header"
  >
    Bydelsfakta – statistikk om befolkning, levekår og boforhold
  </h1>
   
  <div
    class="navigation-topbar navigation-topbar--hidden"
  >
    <!---->
     
    <!---->
  </div>
</header>
`;
